python/pylibraft/pylibraft/neighbors/ivf_pq.pyx

```diff
@@ -0,0 +1,547 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# cython: profile=False
+# distutils: language = c++
+# cython: embedsignature = True
+# cython: language_level = 3
+
+import numpy as np
+import pylibraft.common.handle
+
+from libc.stdint cimport uintptr_t
+from libc.stdint cimport uint32_t, uint8_t, int8_t, int64_t, uint64_t
+from cython.operator cimport dereference as deref
+from libcpp cimport bool, nullptr
+
+from pylibraft.distance.distance_type cimport DistanceType
+from pylibraft.common import Handle
+from pylibraft.common.handle cimport handle_t
+from rmm._lib.memory_resource cimport device_memory_resource
+
+cimport pylibraft.neighbors.c_ivf_pq as c_ivf_pq
+
+from pylibraft.neighbors.c_ivf_pq cimport index_params
+from pylibraft.neighbors.c_ivf_pq cimport search_params
+
+def is_c_cont(cai):
+    dt = np.dtype(cai["typestr"])
+    return "strides" not in cai or \
+        cai["strides"] is None or \
+        cai["strides"][1] == dt.itemsize
+
+
+def _get_metric(metric):
+    SUPPORTED_DISTANCES = {
+        "l2_expanded": DistanceType.L2Expanded,
+        # TODO(tfeher): fix inconsistency: index building for L2SqrtExpanded is only supported by build, not by search.
+        # "euclidean": DistanceType.L2SqrtExpanded
+        "inner_product": DistanceType.InnerProduct
+    }
+    if metric not in SUPPORTED_DISTANCES:
+        raise ValueError("metric %s is not supported" % metric)
+    return SUPPORTED_DISTANCES[metric]
+
+
+def _check_input_array(cai, exp_dt, exp_rows=None, exp_cols=None):
+        if cai["typestr"] not in exp_dt:
+            raise TypeError("dtype %s not supported" % cai["typestr"])
+
+        if not is_c_cont(cai):
+            raise ValueError("Row major input is expected")
+
+        if exp_cols is not None and cai["shape"][1] != exp_cols:
+            raise ValueError("Incorrect number of columns, expected {} got {}" \
+                                 .format(exp_cols, cai["shape"][1]))
+
+        if exp_rows is not None and cai["shape"][0] != exp_rows:
+            raise ValueError("Incorrect number of rows, expected {} , got {}" \
+                                .format(exp_rows, cai["shape"][0]))
+
+
+# Variables to provide easier access for parameters
+PER_SUBSPACE = c_ivf_pq.codebook_gen.PER_SUBSPACE
+PER_CLUSTER = c_ivf_pq.codebook_gen.PER_CLUSTER
+
+CUDA_R_32F = c_ivf_pq.cudaDataType_t.CUDA_R_32F
+CUDA_R_16F = c_ivf_pq.cudaDataType_t.CUDA_R_16F
+CUDA_R_8U = c_ivf_pq.cudaDataType_t.CUDA_R_8U
+
+
+cdef class IndexParams:
+    cdef c_ivf_pq.index_params params
+
+    def __init__(self, *, 
+                 n_lists=1024, 
+                 metric="l2_expanded",
+                 kmeans_n_iters=20, 
+                 kmeans_trainset_fraction=0.5,
+                 pq_bits=8,
+                 pq_dim=0,
+                 codebook_kind="per_subspace",
```

Comment:
You are right, "subspace" or "cluster" is simpler. I had the general question, whether to use strings, or the enums which we are also exporting eg `ivf_pq.PER_SUBSPACE`. The same question applies for the metric parameter. I will go with the strings for now, since that feels more user friendly. Let me know if you would prefer to have the enums.